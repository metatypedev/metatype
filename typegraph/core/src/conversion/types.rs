// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use common::typegraph::{PolicyIndices, TypeNode, TypeNodeBase};
use enum_dispatch::enum_dispatch;
use indexmap::IndexMap;
use std::rc::Rc;

use crate::errors::Result;
use crate::typegraph::TypegraphContext;

#[enum_dispatch]
pub trait TypeConversion {
    /// takes already converted runtime id
    fn convert(&self, ctx: &mut TypegraphContext, runtime_id: Option<u32>) -> Result<TypeNode>;
}

impl<T: TypeConversion> TypeConversion for Rc<T> {
    fn convert(&self, ctx: &mut TypegraphContext, runtime_id: Option<u32>) -> Result<TypeNode> {
        (**self).convert(ctx, runtime_id)
    }
}

#[derive(Default)]
pub struct TypeNodeBaseBuilder {
    name: String,
    runtime: u32,
    enumeration: Option<Vec<String>>,
    policies: Vec<PolicyIndices>,
    runtime_config: Option<Vec<(String, String)>>,
    as_id: bool,
}

/// takes converted runtime id
pub fn gen_base(
    name: String,
    runtime_config: Option<Vec<(String, String)>>,
    runtime_id: u32,
) -> TypeNodeBaseBuilder {
    TypeNodeBaseBuilder {
        name,
        runtime: runtime_id,
        runtime_config,
        ..Default::default()
    }
}

impl TypeNodeBaseBuilder {
    pub fn build(self) -> TypeNodeBase {
        let config = self.runtime_config.map(|c| {
            c.iter()
                .map(|(k, v)| (k.to_string(), serde_json::from_str(v).unwrap()))
                .collect::<IndexMap<_, _>>()
        });

        TypeNodeBase {
            config: config.unwrap_or(Default::default()),
            description: None,
            enumeration: self.enumeration,
            injection: None,
            policies: self.policies,
            runtime: self.runtime,
            title: self.name,
            as_id: self.as_id,
        }
    }

    pub fn enum_(mut self, enumeration: Option<Vec<String>>) -> Self {
        self.enumeration = enumeration;
        self
    }

    pub fn id(mut self, b: bool) -> Self {
        self.as_id = b;
        self
    }
}
