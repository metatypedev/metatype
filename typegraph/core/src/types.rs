use enum_dispatch::enum_dispatch;
use serde::Serialize;
use std::collections::HashMap;

use crate::core::Tpe;

#[enum_dispatch]
pub trait TypeFun {
    fn getattr(&self, field: String) -> Option<Tpe>;
}

#[derive(Debug)]
#[enum_dispatch(TypeFun)]
pub enum T {
    Struct(Struct),
    Integer(Integer),
}

#[derive(Debug, Default, Serialize)]
pub struct Struct {
    pub props: HashMap<String, u32>,
}

impl TypeFun for Struct {
    fn getattr(&self, field: String) -> Option<Tpe> {
        self.props.get(&field).map(|x| Tpe { id: *x })
    }
}

#[derive(Debug, Default, Serialize, Clone)]
pub struct Integer {
    pub min: Option<i32>,
    pub max: Option<i32>,
}

impl TypeFun for Integer {
    fn getattr(&self, _field: String) -> Option<Tpe> {
        None
    }
}
