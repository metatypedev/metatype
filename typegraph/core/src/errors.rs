// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::wit::core::Error as TgError;

pub type Result<T, E = TgError> = std::result::Result<T, E>;

pub fn invalid_max_value() -> TgError {
    "min must be less than or equal to max".to_string()
}

pub fn duplicate_key(name: &str) -> TgError {
    format!("duplicate key '{name}' in properties")
}

pub fn invalid_prop_key(name: &str) -> TgError {
    format!(
        "'{name}' is not a valid property key: allowed characters are ascii letters and underscores",
    )
}

// pub fn property_not_found_in(ty: &str, prop_name: &str) -> TgError {
//     format!("property '{prop_name}' not found in type '{ty}'")
// }

pub fn invalid_input_type(got: &str) -> TgError {
    format!("expected a Struct as input type but got {got}")
}

pub fn invalid_type(expected: &str, got: &str) -> TgError {
    format!("expected {expected} but got {got}")
}

pub fn nested_typegraph_context(active: &str) -> TgError {
    format!("cannot init typegraph: typegraph '{active}' is still active")
}

pub fn expected_typegraph_context() -> TgError {
    "no active typegraph context".to_string()
}

pub fn invalid_export_type(name: &str, got: &str) -> TgError {
    format!("expected a Func to be exposed, got {got} under the name '{name}'")
}

pub fn invalid_export_name(name: &str) -> TgError {
    format!("invalid export name '{name}': allowed characters are ascii letters and underscores")
}

pub fn duplicate_export_name(name: &str) -> TgError {
    format!("duplicate export name '{name}'")
}

pub fn unregistered_type_name(name: &str) -> TgError {
    format!("type name '{name}' has not been registered")
}

pub fn object_not_found(kind: &str, id: u32) -> TgError {
    format!("{kind} #{id} not found")
}

pub fn invalid_path(pos: usize, path: &[String], curr_keys: &[String]) -> TgError {
    let mut path_with_cursor = vec![];
    for (i, chunk) in path.iter().enumerate() {
        if i == pos {
            path_with_cursor.push(format!("[{}]", chunk));
        } else {
            path_with_cursor.push(chunk.clone());
        }
    }
    format!(
        "invalid path {:?}, none of {} match the chunk {:?}",
        path_with_cursor.join("."),
        curr_keys.join(", "),
        path.get(pos).unwrap_or(&"".to_string()),
    )
}

pub fn expect_object_at_path(path: &[String]) -> TgError {
    format!("object was expected at path {:?}", path.join("."))
}

pub fn unknown_predefined_function(name: &str, runtime: &str) -> TgError {
    format!("unknown predefined function {name} for runtime {runtime}")
}

pub fn duplicate_policy_name(name: &str) -> TgError {
    format!("duplicate policy name '{name}'")
}

pub fn base_required(name: &str) -> TgError {
    format!("Concrete type '{name}' must have a base.")
}

pub fn base_not_allowed(name: &str) -> TgError {
    format!("Wrapper type '{name}' must not have a base.")
}
