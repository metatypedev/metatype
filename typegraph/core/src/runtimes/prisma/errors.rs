// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::wit::core::Error;

// pub fn relationship_not_found(source_model: &str, field: &str) -> Error {
//     format!("relationship target not found for  {source_model}::{field}")
// }

pub fn ambiguous_side(
    first_model: &str,
    first_field: &str,
    second_model: &str,
    second_field: &str,
) -> Error {
    [
        "Ambiguous side:",
        "cannot determine which model should have the foreign key:",
        &format!(
            "{}::{} or {}::{}",
            first_model, first_field, second_model, second_field
        ),
        "Please add 'fkey' or 'unique' attribute on one side",
    ]
    .join(" ")
    .into()
}

pub fn conflicting_attributes(
    attribute: &str,
    first_model: &str,
    first_field: &str,
    second_model: &str,
    second_field: &str,
) -> Error {
    [
        format!("Conflicting attributes '{attribute}':"),
        format!("on {first_model}::{first_field}"),
        format!("and {second_model}::{second_field}"),
    ]
    .join(" ")
    .into()
}

pub fn no_relationship_target(model: &str, field: &str, target_model: &str) -> Error {
    format!(r#"Relationship target field not found for "{model}::{field}" on {target_model:?}."#)
        .into()
}
