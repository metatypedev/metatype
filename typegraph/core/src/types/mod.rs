// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

pub mod subgraph;
pub mod type_def;
pub mod type_id;
pub mod type_ref;

pub use type_def::*;
pub use type_id::*;
pub use type_ref::*;

#[derive(Clone, Debug)]
pub enum Type {
    Ref(TypeRef),
    Def(TypeDef),
}

impl From<TypeRef> for Type {
    fn from(r: TypeRef) -> Self {
        Self::Ref(r)
    }
}

impl From<TypeDef> for Type {
    fn from(d: TypeDef) -> Self {
        Self::Def(d)
    }
}

impl Type {
    fn name(&self) -> Option<&str> {
        match self {
            Type::Ref(typ) => Some(typ.name.as_str()),
            Type::Def(typ) => typ.name(),
        }
    }

    fn repr(&self) -> String {
        match self {
            Type::Ref(typ) => typ.repr(),
            Type::Def(typ) => typ.repr(),
        }
    }
}
