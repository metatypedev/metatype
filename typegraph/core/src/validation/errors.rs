// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::wit::core::Error;

// pub fn invalid_runtime_type(runtime: &str, materializer: &str) -> Error {
//     format!(
//         "invalid runtime type {} for materializer {}",
//         runtime, materializer
//     )
// }

pub fn invalid_output_type_predefined(name: &str, expected: &str, got: &str) -> Error {
    format!(
        "invalid output type for predefined function {}: expected {}, got {}",
        name, expected, got
    )
    .into()
}

pub fn unknown_predefined_function(name: &str) -> Error {
    format!("unknown predefined function {}", name).into()
}
