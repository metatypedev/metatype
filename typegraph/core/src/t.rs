// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::Result;
use crate::types::TypeId;
use crate::wit::core::{
    Core, TypeArray, TypeBase, TypeEither, TypeFloat, TypeFunc, TypeInteger, TypeOptional,
    TypeProxy, TypeString, TypeStruct, TypeUnion,
};

pub trait TypeBuilder {
    fn build(&mut self) -> Result<TypeId>;
}

pub trait ConcreteTypeBuilder: TypeBuilder {
    fn base_mut(&mut self) -> &mut TypeBase;

    fn named(&mut self, name: impl Into<String>) -> &mut Self {
        self.base_mut().name = Some(name.into());
        self
    }

    fn as_id(&mut self, as_id: bool) -> &mut Self {
        self.base_mut().as_id = as_id;
        self
    }

    fn config(&mut self, key: impl Into<String>, value: impl Into<String>) -> &mut Self {
        let runtime_config = &mut self.base_mut().runtime_config;
        if runtime_config.is_none() {
            *runtime_config = Some(Default::default());
        }
        runtime_config
            .as_mut()
            .unwrap()
            .push((key.into(), value.into()));
        self
    }
}

#[derive(Default)]
pub struct BooleanBuilder {
    base: TypeBase,
}

pub fn boolean() -> BooleanBuilder {
    Default::default()
}

#[derive(Default)]
pub struct IntegerBuilder {
    base: TypeBase,
    data: TypeInteger,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeInteger {
    fn default() -> Self {
        Self {
            min: None,
            max: None,
            exclusive_minimum: None,
            exclusive_maximum: None,
            multiple_of: None,
            enumeration: None,
        }
    }
}

impl IntegerBuilder {
    #[allow(dead_code)]
    pub fn min(mut self, min: i32) -> Self {
        self.data.min = Some(min);
        self
    }

    #[allow(dead_code)]
    pub fn max(mut self, max: i32) -> Self {
        self.data.max = Some(max);
        self
    }

    #[allow(dead_code)]
    pub fn x_min(mut self, min: i32) -> Self {
        self.data.exclusive_minimum = Some(min);
        self
    }

    #[allow(dead_code)]
    pub fn x_max(mut self, max: i32) -> Self {
        self.data.exclusive_maximum = Some(max);
        self
    }
}

pub fn integer() -> IntegerBuilder {
    Default::default()
}

#[derive(Default)]
pub struct FloatBuilder {
    base: TypeBase,
    data: TypeFloat,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeFloat {
    fn default() -> Self {
        Self {
            min: None,
            max: None,
            exclusive_minimum: None,
            exclusive_maximum: None,
            multiple_of: None,
            enumeration: None,
        }
    }
}

impl FloatBuilder {
    #[allow(dead_code)]
    pub fn min(mut self, min: f64) -> Self {
        self.data.min = Some(min);
        self
    }

    #[allow(dead_code)]
    pub fn max(mut self, max: f64) -> Self {
        self.data.max = Some(max);
        self
    }

    #[allow(dead_code)]
    pub fn x_min(mut self, min: f64) -> Self {
        self.data.exclusive_minimum = Some(min);
        self
    }

    #[allow(dead_code)]
    pub fn x_max(mut self, max: f64) -> Self {
        self.data.exclusive_maximum = Some(max);
        self
    }
}

pub fn float() -> FloatBuilder {
    Default::default()
}

#[derive(Default)]
pub struct StringBuilder {
    base: TypeBase,
    data: TypeString,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeString {
    fn default() -> Self {
        Self {
            min: None,
            max: None,
            format: None,
            pattern: None,
            enumeration: None,
        }
    }
}

pub fn string() -> StringBuilder {
    Default::default()
}

impl StringBuilder {
    #[allow(dead_code)]
    pub fn format(&mut self, format: impl Into<String>) -> &mut Self {
        self.data.format = Some(format.into());
        self
    }

    pub fn enum_(&mut self, values: Vec<String>) -> &mut Self {
        self.data.enumeration = Some(
            values
                .into_iter()
                .map(|v| serde_json::to_string(&serde_json::Value::String(v)).unwrap())
                .collect(),
        );
        self
    }
}

#[derive(Default)]
pub struct OptionalBuilder {
    base: TypeBase,
    data: TypeOptional,
}

impl Default for TypeOptional {
    fn default() -> Self {
        Self {
            of: u32::max_value(),
            default_item: None,
        }
    }
}

pub fn optional(ty: TypeId) -> OptionalBuilder {
    OptionalBuilder {
        base: TypeBase::default(),
        data: TypeOptional {
            of: ty.into(),
            default_item: None,
        },
    }
}

#[derive(Default)]
pub struct ArrayBuilder {
    base: TypeBase,
    data: TypeArray,
}

impl Default for TypeArray {
    fn default() -> Self {
        Self {
            of: u32::max_value(),
            min: None,
            max: None,
            unique_items: None,
        }
    }
}

pub fn array(ty: TypeId) -> ArrayBuilder {
    ArrayBuilder {
        base: TypeBase::default(),
        data: TypeArray {
            of: ty.into(),
            ..Default::default()
        },
    }
}

#[derive(Default)]
pub struct UnionBuilder {
    base: TypeBase,
    data: TypeUnion,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeUnion {
    fn default() -> Self {
        Self {
            variants: Default::default(),
        }
    }
}

pub fn union(variants: impl IntoIterator<Item = TypeId>) -> UnionBuilder {
    UnionBuilder {
        data: TypeUnion {
            variants: variants.into_iter().map(|tid| tid.0).collect(),
        },
        ..Default::default()
    }
}

#[derive(Default)]
pub struct EitherBuilder {
    base: TypeBase,
    data: TypeEither,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeEither {
    fn default() -> Self {
        Self {
            variants: Default::default(),
        }
    }
}

pub fn either(variants: impl IntoIterator<Item = TypeId>) -> EitherBuilder {
    EitherBuilder {
        data: TypeEither {
            variants: variants.into_iter().map(|tid| tid.0).collect(),
        },
        ..Default::default()
    }
}

#[derive(Default)]
pub struct StructBuilder {
    base: TypeBase,
    data: TypeStruct,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeStruct {
    fn default() -> Self {
        Self {
            props: Vec::new(),
            additional_props: false,
            min: None,
            max: None,
        }
    }
}

pub fn struct_() -> StructBuilder {
    Default::default()
}

pub fn struct_from(props: impl Iterator<Item = (String, TypeId)>) -> StructBuilder {
    StructBuilder {
        data: TypeStruct {
            props: props.map(|(k, ty)| (k, ty.into())).collect(),
            ..Default::default()
        },
        ..Default::default()
    }
}

pub fn struct_extends(ty: TypeId) -> Result<StructBuilder> {
    Ok(StructBuilder {
        data: TypeStruct {
            props: ty.as_struct().map(|typ| typ.data.props.clone())?,
            ..Default::default()
        },
        ..Default::default()
    })
}

impl StructBuilder {
    pub fn prop(&mut self, name: impl Into<String>, ty: TypeId) -> &mut Self {
        self.data.props.push((name.into(), ty.into()));
        self
    }

    #[allow(dead_code)]
    pub fn props(&mut self, props: impl IntoIterator<Item = (String, TypeId)>) {
        self.data
            .props
            .extend(props.into_iter().map(|(name, ty)| (name, ty.into())));
    }

    pub fn min(&mut self, min: u32) -> &mut Self {
        self.data.min = Some(min);
        self
    }

    pub fn max(&mut self, max: u32) -> &mut Self {
        self.data.max = Some(max);
        self
    }
}

#[derive(Default)]
pub struct FuncBuilder {
    #[allow(dead_code)]
    base: TypeBase,
    data: TypeFunc,
}

impl Default for TypeFunc {
    fn default() -> Self {
        Self {
            inp: u32::max_value(),
            out: u32::max_value(),
            mat: u32::max_value(),
        }
    }
}

#[allow(dead_code)]
pub fn func(inp: TypeId, out: TypeId, mat: u32) -> Result<TypeId> {
    FuncBuilder {
        data: TypeFunc {
            inp: inp.into(),
            out: out.into(),
            mat,
        },
        ..Default::default()
    }
    .build()
}

#[derive(Default)]
pub struct ProxyBuilder {
    data: TypeProxy,
}

impl Default for TypeProxy {
    fn default() -> Self {
        Self {
            name: "".to_string(),
            extras: vec![],
        }
    }
}

impl ProxyBuilder {
    /// Adds extra data entry in the proxy
    pub fn set(&mut self, key: impl Into<String>, value: impl Into<String>) -> &mut Self {
        self.data.extras.push((key.into(), value.into()));
        self
    }
}

pub fn proxy(name: impl Into<String>) -> ProxyBuilder {
    ProxyBuilder {
        data: TypeProxy {
            name: name.into(),
            ..Default::default()
        },
    }
}

macro_rules! impl_type_builder {
    ( $ty:ty, $build:ident ) => {
        impl TypeBuilder for $ty {
            fn build(&mut self) -> Result<TypeId> {
                let builder = std::mem::take(self);
                Ok($crate::Lib::$build(builder.data, builder.base)?.into())
            }
        }

        impl ConcreteTypeBuilder for $ty {
            fn base_mut(&mut self) -> &mut TypeBase {
                &mut self.base
            }
        }
    };

    ( $ty:ty, $build:ident, true ) => {
        impl TypeBuilder for $ty {
            fn build(&mut self) -> Result<TypeId> {
                let builder = std::mem::take(self);
                Ok($crate::Lib::$build(builder.data)?.into())
            }
        }
    };
}
impl TypeBuilder for BooleanBuilder {
    fn build(&mut self) -> Result<TypeId> {
        let builder = std::mem::take(self);
        Ok(crate::Lib::booleanb(builder.base)?.into())
    }
}

impl ConcreteTypeBuilder for BooleanBuilder {
    fn base_mut(&mut self) -> &mut TypeBase {
        &mut self.base
    }
}

impl_type_builder!(IntegerBuilder, integerb);
impl_type_builder!(FloatBuilder, floatb);
impl_type_builder!(OptionalBuilder, optionalb);
impl_type_builder!(StringBuilder, stringb);
impl_type_builder!(ArrayBuilder, arrayb);
impl_type_builder!(UnionBuilder, unionb);
impl_type_builder!(EitherBuilder, eitherb);
impl_type_builder!(StructBuilder, structb);
impl_type_builder!(FuncBuilder, funcb, true);
impl_type_builder!(ProxyBuilder, proxyb, true);
