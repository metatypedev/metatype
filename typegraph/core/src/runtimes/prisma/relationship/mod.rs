// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::Result;
use crate::t;
use crate::t::TypeBuilder;
use crate::types::TypeId;

mod discovery;
pub mod registry;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Cardinality {
    Optional,
    One,
    Many,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct RelationshipModel {
    pub model_type: TypeId,
    pub model_name: String,
    pub wrapper_type: TypeId,
    pub cardinality: Cardinality,
    pub field: String,
}

#[derive(Debug, Clone, Copy)]
pub enum Side {
    Left,
    Right,
}

impl Side {
    #[allow(dead_code)]
    pub fn opposite(&self) -> Self {
        match self {
            Side::Left => Side::Right,
            Side::Right => Side::Left,
        }
    }
}

/// Possible cardinalities are:
/// (Optional, Optional): [Left] 0..1 --> 0..1 [Right]
/// (One, Optional): [Left] 1..1 --> 0..1 [Right]
/// (Optional, Many) [Left] 0..1 --> 0..n [Right]
/// (One, Many) [Left] 1..1 --> 0..n [Right]
/// The model on the right will have the foreign key
#[derive(Debug, Clone)]
pub struct Relationship {
    pub name: String,
    pub left: RelationshipModel,
    pub right: RelationshipModel,
}

pub enum SideOfModel {
    Left,
    Right,
    Both,
    None,
}

impl Relationship {
    pub fn get_opposite_of(&self, model_id: TypeId, field: &str) -> Option<&RelationshipModel> {
        use SideOfModel as S;
        match self.side_of_model(model_id) {
            S::Both => {
                if self.left.field == field {
                    Some(&self.right)
                } else if self.right.field == field {
                    Some(&self.left)
                } else {
                    unreachable!()
                }
            }
            S::Left => Some(&self.right),
            S::Right => Some(&self.left),
            S::None => None,
        }
    }

    fn side_of_model(&self, model_type: TypeId) -> SideOfModel {
        use SideOfModel as S;
        if self.left.model_type == self.right.model_type {
            if self.left.model_type == model_type {
                S::Both
            } else {
                S::None
            }
        } else if self.left.model_type == model_type {
            S::Left
        } else if self.right.model_type == model_type {
            S::Right
        } else {
            S::None
        }
    }

    pub fn side_of_type(&self, type_id: TypeId) -> Option<Side> {
        if self.left.wrapper_type == type_id {
            Some(Side::Left)
        } else if self.right.wrapper_type == type_id {
            Some(Side::Right)
        } else {
            None
        }
    }

    pub fn get(&self, side: Side) -> &RelationshipModel {
        match side {
            Side::Left => &self.left,
            Side::Right => &self.right,
        }
    }
}

#[derive(Default, Clone)]
pub struct PrismaLink {
    type_name: String,
    rel_name: Option<String>,
    fkey: Option<bool>,
    target_field: Option<String>,
    unique: bool,
}

impl PrismaLink {
    pub fn name(mut self, n: impl Into<String>) -> Self {
        self.rel_name = Some(n.into());
        self
    }

    pub fn fkey(mut self, fk: bool) -> Self {
        self.fkey = Some(fk);
        self
    }

    pub fn field(mut self, field: impl Into<String>) -> Self {
        self.target_field = Some(field.into());
        self
    }

    pub fn unique(mut self, unique: bool) -> Self {
        self.unique = unique;
        self
    }

    fn build_link(&self) -> Result<TypeId> {
        let mut proxy = t::proxy(&self.type_name);
        if let Some(rel_name) = self.rel_name.clone() {
            proxy.set("rel_name", rel_name);
        }
        if let Some(fkey) = self.fkey {
            proxy.set("fkey", format!("{fkey}"));
        }
        if let Some(target_field) = self.target_field.clone() {
            proxy.set("target_field", target_field);
        }
        let res = proxy.build()?;
        eprintln!("proxy: {:?}", res);
        Ok(res)
    }
}

impl TypeBuilder for PrismaLink {
    fn build(&self) -> Result<TypeId> {
        self.build_link()
    }
}

#[allow(dead_code)]
pub fn prisma_linkx(typ: impl TypeBuilder) -> Result<PrismaLink> {
    prisma_link(typ.build()?)
}

pub fn prisma_link(type_id: TypeId) -> Result<PrismaLink> {
    let name = type_id
        .type_name()?
        .ok_or_else(|| "Prisma link target must be named".to_string())?;
    Ok(prisma_linkn(name))
}

pub fn prisma_linkn(name: impl Into<String>) -> PrismaLink {
    PrismaLink {
        type_name: name.into(),
        ..Default::default()
    }
}

use registry::RelationshipRegistry;

#[cfg(test)]
mod test {
    use super::{prisma_linkn, prisma_linkx};
    use crate::errors::Result;
    use crate::global_store::Store;
    use crate::runtimes::prisma::errors;
    use crate::runtimes::prisma::relationship::registry::RelationshipRegistry;
    use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
    use crate::test_utils::*;

    #[test]
    fn test_implicit_relationships() -> Result<()> {
        let (user, _post) = models::simple_relationship()?;

        let mut reg = RelationshipRegistry::default();
        reg.manage(user)?;

        insta::assert_debug_snapshot!("implicit relationship", reg);

        Ok(())
    }

    #[test]
    fn test_explicit_relationship_name() -> Result<()> {
        Store::reset();
        let user = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx("name", t::string())?
            .propx("posts", t::arrayx(t::proxy("Post"))?)?
            .named("User")
            .build()?;

        let post = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx("title", t::string())?
            .propx("author", prisma_linkn("User").name("PostAuthor"))?
            .named("Post")
            .build()?;

        let mut reg = RelationshipRegistry::default();
        reg.manage(user)?;
        reg.manage(post)?;

        insta::assert_debug_snapshot!("explicitly named relationship", reg);

        Ok(())
    }

    #[test]
    fn test_fkey_attribute() -> Result<()> {
        Store::reset();
        let user = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx(
                "profile",
                prisma_linkx(t::optionalx(t::proxy("Profile"))?)?.fkey(true),
            )?
            .named("User")
            .build()?;

        let profile = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx("user", t::optionalx(t::proxy("User"))?)?
            .named("Profile")
            .build()?;

        let mut reg = RelationshipRegistry::default();
        reg.manage(user)?;
        reg.manage(profile)?;

        insta::assert_debug_snapshot!("fkey attribute", reg);

        Ok(())
    }

    #[test]
    fn test_unique_attribute() -> Result<()> {
        Store::reset();
        let user = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx(
                "profile",
                t::optionalx(t::proxy("Profile"))?.config("unique", "true"),
            )?
            .named("User")
            .build()?;

        let profile = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx("user", t::optionalx(t::proxy("User"))?)?
            .named("Profile")
            .build()?;

        let mut reg = RelationshipRegistry::default();
        reg.manage(user)?;
        reg.manage(profile)?;

        insta::assert_debug_snapshot!("unique attribute", reg);

        Ok(())
    }

    #[test]
    fn test_self_relationship() -> Result<()> {
        Store::reset();
        let node = t::struct_()
            .propx("id", t::string().as_id(true))?
            .propx("children", t::arrayx(t::proxy("Node"))?)?
            .propx("parent", t::proxy("Node"))?
            .named("Node")
            .build()?;

        let mut reg = RelationshipRegistry::default();
        reg.manage(node)?;

        insta::assert_debug_snapshot!("self relationship", reg);

        Ok(())
    }

    #[test]
    fn test_ambiguous_side() -> Result<()> {
        Store::reset();
        let user = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx("profile", t::proxy("Profile"))?
            .named("User")
            .build()?;

        let profile = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx("user", t::proxy("User"))?
            .named("Profile")
            .build()?;

        let mut reg = RelationshipRegistry::default();
        let res = reg.manage(user);
        assert_eq!(
            res,
            Err(errors::ambiguous_side("Profile", "user", "User", "profile"))
        );
        let res = reg.manage(profile);
        assert_eq!(
            res,
            Err(errors::ambiguous_side("User", "profile", "Profile", "user"))
        );

        Store::reset();
        let user = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx("profile", t::optionalx(t::proxy("Profile"))?)?
            .named("User")
            .build()?;

        let profile = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx("user", t::optionalx(t::proxy("User"))?)?
            .named("Profile")
            .build()?;

        let mut reg = RelationshipRegistry::default();
        let res = reg.manage(user);
        assert_eq!(
            res,
            Err(errors::ambiguous_side("Profile", "user", "User", "profile"))
        );
        let res = reg.manage(profile);
        assert_eq!(
            res,
            Err(errors::ambiguous_side("User", "profile", "Profile", "user"))
        );

        Ok(())
    }

    #[test]
    fn test_conflicting_attributes() -> Result<()> {
        Store::reset();
        let user = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx("profile", prisma_linkn("Profile").fkey(true))?
            .named("User")
            .build()?;

        let profile = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx("user", prisma_linkn("User").fkey(true))?
            .named("Profile")
            .build()?;

        let mut reg = RelationshipRegistry::default();
        let res = reg.manage(user);
        assert_eq!(
            res,
            Err(errors::conflicting_attributes(
                "fkey", "Profile", "user", "User", "profile"
            ))
        );
        let res = reg.manage(profile);
        assert_eq!(
            res,
            Err(errors::conflicting_attributes(
                "fkey", "User", "profile", "Profile", "user"
            ))
        );

        Store::reset();
        let user = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx("profile", prisma_linkn("Profile").fkey(false))?
            .named("User")
            .build()?;

        let profile = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx("user", prisma_linkn("User").fkey(false))?
            .named("Profile")
            .build()?;

        let mut reg = RelationshipRegistry::default();
        let res = reg.manage(user);
        assert_eq!(
            res,
            Err(errors::conflicting_attributes(
                "fkey", "Profile", "user", "User", "profile"
            ))
        );
        let res = reg.manage(profile);
        assert_eq!(
            res,
            Err(errors::conflicting_attributes(
                "fkey", "User", "profile", "Profile", "user"
            ))
        );

        Ok(())
    }

    #[test]
    fn test_missing_target() -> Result<()> {
        Store::reset();
        let user = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .propx("profile", prisma_linkn("Profile").fkey(true))?
            .named("User")
            .build()?;

        let _profile = t::struct_()
            .propx("id", t::integer().as_id(true))?
            .named("Profile")
            .build()?;

        let mut reg = RelationshipRegistry::default();
        let res = reg.manage(user);
        assert_eq!(
            res,
            Err(errors::no_relationship_target("User", "profile", "Profile"))
        );

        Ok(())
    }
}
