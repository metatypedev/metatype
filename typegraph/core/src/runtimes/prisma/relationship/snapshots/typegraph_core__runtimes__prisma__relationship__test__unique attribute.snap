---
source: typegraph/core/src/runtimes/prisma/relationship/mod.rs
expression: reg
---
RelationshipRegistry {
    models: {
        Type#3: RegisteredModel {
            relationships: {
                "profile": "__rel_User_Profile_1",
            },
            name: "User",
            id_field: "id",
        },
        Type#7: RegisteredModel {
            relationships: {
                "user": "__rel_User_Profile_1",
            },
            name: "Profile",
            id_field: "id",
        },
    },
    models_by_name: {},
    relationships: {
        "__rel_User_Profile_1": Relationship {
            name: "__rel_User_Profile_1",
            left: RelationshipModel {
                model_type: Type#7,
                model_name: "Profile",
                wrapper_type: Type#2,
                cardinality: Optional,
                field: "user",
            },
            right: RelationshipModel {
                model_type: Type#3,
                model_name: "User",
                wrapper_type: Type#6,
                cardinality: Optional,
                field: "profile",
            },
        },
    },
    complete_registrations: {
        Type#3,
        Type#7,
    },
    counter: RefCell {
        value: 1,
    },
}
